### PR TITLE
Try tabs instead of segmented control for switching between solid/gradient in color panels

### PR DESCRIPTION
**Pull request by jameskoster**
When picking a color, where both solid colors and gradients are available, their visibility is toggled through a component that resembles a segmented control:

Screenshot 2022-06-23 at 10 28 17

One drawback I find with this approach is that it is not a consistent implementation of the segmented control pattern when compared with other instances in our UI. When choosing a font size (for example) a selection equates with an actual value, whereas with colors the value isn't set until a color swatch is clicked.

To promote consistency we might instead use tabs for switching between the solid / gradient palettes:

https://user-images.githubusercontent.com/846565/175508545-b541b6ad-cc94-41fe-9b0c-4e81c1e8cb60.mp4

This feels like an appropriate use case for the TabPanel component:

> TabPanels organize and allow navigation between groups of content that are related and at the same level of hierarchy.
> – [Block Editor Handbook](https://developer.wordpress.org/block-editor/reference-guides/components/tab-panel/)

---

A somewhat related consideration is the tab labels. Since they toggle the visibility of different palettes, and palettes typically contain more than one color, I wonder if the labels should be plural?

Screenshot 2022-06-24 at 10 38 44

This is probably something we could adjust here as well, if it is deemed appropriate.

### CODE REVIEW

File: packages/block-editor/src/components/colors-gradients/control.js
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/control.js]: This may be a bit of an early optimization in the PR, but we could extract the values of the `tabs` to a separate variables, in order to avoid re-creating the same object every render (and to re-use the variable across the component instead of hardcoding the tab IDs around the code).

<details>

<summary>For example</summary>

```diff
diff --git a/packages/block-editor/src/components/colors-gradients/control.js b/packages/block-editor/src/components/colors-gradients/control.js
index 8d38e8e901..f58be3dc82 100644
--- a/packages/block-editor/src/components/colors-gradients/control.js
+++ b/packages/block-editor/src/components/colors-gradients/control.js
@@ -29,6 +29,19 @@ const colorsAndGradientKeys = [
 	'disableCustomGradients',
 ];
 
+const TAB_COLOR = {
+	name: 'color',
+	title: 'Solid color',
+	value: 'color',
+};
+const TAB_GRADIENT = {
+	name: 'gradient',
+	title: 'Gradient',
+	value: 'gradient',
+};
+
+const TABS_SETTINGS = [ TAB_COLOR, TAB_GRADIENT ];
+
 function ColorGradientControlInner( {
 	colors,
 	gradients,
@@ -52,7 +65,9 @@ function ColorGradientControlInner( {
 		onGradientChange &&
 		( ! isEmpty( gradients ) || ! disableCustomGradients );
 	const [ currentTab, setCurrentTab ] = useState(
-		gradientValue ? 'gradient' : !! canChooseAColor && 'color'
+		gradientValue
+			? TAB_GRADIENT.value
+			: !! canChooseAColor && TAB_COLOR.value
 	);
 
 	if ( ! canChooseAColor && ! canChooseAGradient ) {
@@ -80,23 +95,13 @@ function ColorGradientControlInner( {
 						<TabPanel
 							className="block-editor-color-gradient-control__tabs"
 							onSelect={ setCurrentTab }
-							tabs={ [
-								{
-									name: 'color',
-									title: 'Solid color',
-									value: 'color',
-								},
-								{
-									name: 'gradient',
-									title: 'Gradient',
-									value: 'gradient',
-								},
-							] }
+							tabs={ TABS_SETTINGS }
 						>
 						{ ( tab ) => <p className="screen-reader-text">Selected tab: { tab.title }</p> }
 						</TabPanel>
 					) }
-					{ ( currentTab === 'color' || ! canChooseAGradient ) && (
+					{ ( currentTab === TAB_COLOR.value ||
+						! canChooseAGradient ) && (
 						<ColorPalette
 							value={ colorValue }
 							onChange={
@@ -118,7 +123,8 @@ function ColorGradientControlInner( {
 							enableAlpha={ enableAlpha }
 						/>
 					) }
-					{ ( currentTab === 'gradient' || ! canChooseAColor ) && (
+					{ ( currentTab === TAB_GRADIENT.value ||
+						! canChooseAColor ) && (
 						<GradientPicker
 							value={ gradientValue }
 							onChange={
```
</details>

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/style.scss]: I wonder if there's a better way to make sure that the Tabs fill the whole width of the sidebar — what about removing the padding from `.edit-site-screen-background-color__control` and adding it back only to the tab panel's contents?

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/style.scss]: Overriding styles this way introduces fragmentation in the way a component looks across the app, and makes this code less resilient to future design changes.

It would be great if we could use the component as it comes out of the box (or alternatively, we could look at updating the component to fit the new design requirements).

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (jameskoster) [reply packages/block-editor/src/components/colors-gradients/style.scss]: I suppose avoiding negative values has some merit, but both implementations would still be exposed to the same risk, no? Just trying to fully understand the implications :)

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (jameskoster) [reply packages/block-editor/src/components/colors-gradients/style.scss]: As far as I know all implementations of `TabPanel` in core append a bottom border, so we should explore adding it to the component, and removing these styles holistically.

File: packages/block-editor/src/components/colors-gradients/control.js
Comment (jameskoster) [reply packages/block-editor/src/components/colors-gradients/control.js]: I'll make these changes 👍

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/style.scss]: In general, there isn't anything wrong with negative margins. In this case, it's more about introducing custom style overrides of the `TabPanel` component (that implicitly relies on the fact that the popover's content has a padding of `$grid-unit-10` px).

Specifically to this case, please ignore my initial suggestion re. removing the padding from  — I mixed up this instance (which is rendered in a popover) with the one below (rendered in the sidebar).

It looks like the popover's contents have an `8px` margin added directly by the `Dropdown` component ([see the code here](https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/components/src/dropdown/style.scss#L7)) (which I'm definitely not a fan of!)

Screenshot 2022-07-05 at 21 13 55

In this specific situation, I'm not sure if there's a clean way to avoid the style override 🤔 

@mirka , any ideas?

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/style.scss]: That sounds like a good plan, at least in principle!

Comment (jameskoster) [reply packages/block-editor/src/components/colors-gradients/style.scss]: From a purely design perspective I find the tabs to look a little awkward with no stroke beneath them. But thinking about backwards compatibility I suppose we could add a prop to the component that creates a 'borderless' appearance?

Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/style.scss]: On one hand, we could add an `isBorderless` prop (with a default value of `true`, so that it doesn't introduce a visual change to the component) and replace all "border style overrides" by setting `isBorderless` to `false`

But I wonder if this is the right approach:
- I'm not a massive fan of props that are very visually descriptive, as they expose an implementation detail of the design (in this case, showing a border or not) which may not make sense in future iterations
- this pattern of adding props and introduce variants to our components is often a way to mask consistency issues

I'd prefer a scenario in which:
- we investigate how the component is used across Gutenberg (and potentially plugins)
- we assess if the component's design needs updating. In this case, relatively to the bottom border, we should assess if there can be only one version of the component (with or without border) — which would be our preferred option — or if it's absolutely necessary to expose both variants.
- we update the component accordingly and remove style overrides across Gutenberg.

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (mirka) [reply packages/block-editor/src/components/colors-gradients/style.scss]: One solution that comes to mind is how the [`Card` component](https://wordpress.github.io/gutenberg/?path=/story/components-card--default) does it. It has a designated `<CardMedia>` component that you put inside `<Card>` when you want to break out of the padding. We could provide a similar subcomponent for `Dropdown`. ("Media" is [not a very discoverable name](https://github.com/WordPress/gutenberg/issues/39830) for this, so ideally we'll give it a more literal name like `<DropdownContentFullWidth>` or something.)

I'm fine with the hacky style override as a temporary thing for this PR to land, and we can make a `DropdownContentFullWidth` in a separate PR.

File: packages/block-editor/src/components/colors-gradients/control.js
Comment (mirka) [inline packages/block-editor/src/components/colors-gradients/control.js]: I believe what needs to happen here is to provide a function that receives the currently selected tab as an argument and render the tab panel for it. Basically the `TabPanel` is going to manage our `currentTab` state from now on. That is how it handles the a11y concerns for us.

Since it looks like we're going to use the tab panel content outside of a TabPanel context as well, perhaps we could do some light refactoring along these lines:

```jsx
const tabPanels = {
  [TAB_COLOR.value]: <ColorPalette foo={bar} />;
  [TAB_GRADIENT.value]: <GradientPicker foo={bar} />;
};
```

```jsx
					{ canChooseAColor && canChooseAGradient && (
						/* Note that we don't need the `onSelect` anymore */
						<TabPanel
							className="block-editor-color-gradient-control__tabs"
							tabs={ TABS_SETTINGS }
						>
						  { ( tab ) => tabPanels[ tab ]; }
						</TabPanel>
					) }
					{ ! canChooseAGradient && tabPanels[ TAB_COLOR.value ] }
					{ ! canChooseAColor && tabPanels[ TAB_GRADIENT.value ] }
```

File: packages/edit-site/src/components/global-styles/screen-color-palette.js
Comment (mirka) [inline packages/edit-site/src/components/global-styles/screen-color-palette.js]: Same thing here — this is where the tab panel content needs to be rendered.

**Review by mirka — COMMENTED** [review]
Very happy that we're going in this direction! 😊

Once we fix up the usage of the TabPanel component here, I think we can merge and then take note of the follow-up tasks to remove the CSS hacks.

File: packages/block-editor/src/components/colors-gradients/control.js
Comment (jameskoster) [reply packages/block-editor/src/components/colors-gradients/control.js]: Oh yeah, that makes sense :)

I tried to implement this as you suggested but my designer brain has obviously messed something up because the panels aren't rendering for me 🤔 I don't suppose you could take a look?

Comment (mirka) [reply packages/block-editor/src/components/colors-gradients/control.js]: Ah yes, of course! I wrote that code snippet without running it so something is likely wrong 😂 I'll rework the code and push it directly to this PR branch then 👍

File: packages/edit-site/src/components/global-styles/screen-link-color.js
Comment (mirka) [inline packages/edit-site/src/components/global-styles/screen-link-color.js:78]: 🧹 Removed some copypasta here

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (mirka) [inline packages/block-editor/src/components/colors-gradients/style.scss]: FWIW, after working on this PR, I'm now a little apprehensive about adding in this ad hoc border. The Global Styles ▸ Colors menu already contains several tabbed color panels that have different implementations, easily leading to style fragmentation. See also #41976 which seems to have landed recently. Your call, but it might be better to leave the border off for now and address the [border issue](https://github.com/WordPress/gutenberg/pull/41937#discussion_r914673254) in a separate PR.

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (mirka) [reply packages/block-editor/src/components/colors-gradients/style.scss]: Looking at the mockups in #38277 (cf. some of the implementations in Global Styles ▸ Colors), I'm wondering whether we've thought through the decision to make the tablist break out of the parent padding or not. There seems to be a lot of hierarchical contexts where it would be better _not_ to be full-width. I wanted to check on this point because it does add a good amount of complexity and fragmentation risk if we don't have a clear style guide.

It's totally fine if some top-level tablists are full-width (e.g. Blocks/Patterns, Template/Block, Post/Block). But for things that aren't really top-level, or for components that are reused in multiple contexts such as this `ColorGradientControl`, things will be a lot simpler if we don't go full-width. Again, your call, but just a consideration from the maintenance/consistency standpoint 🙂 (cc @javierarce)

File: packages/block-editor/src/components/colors-gradients/control.js
Comment (mirka) [reply packages/block-editor/src/components/colors-gradients/control.js]: Ok, done ✅ Passing the torch back to you!

**Comment by jameskoster** [comment]
I've removed the ad hoc border, and the negative margins in the popover. I agree it would be best to address this at the component level first.

**Review by mirka — APPROVED** [review]
Thanks! I think this is a solid accessibility and consistency improvement 💛 I'll see what we can do about the dropdown paddings. And for the tablist bottom border, let us know which way you want to go and we can help implement it 👍

**Comment by jameskoster** [comment]
Thanks for the review :) I'll merge once another designer has signed off.

> And for the tablist bottom border, let us know which way you want to go and we can help implement it

I'll open an issue for us to discuss this in more detail.

**Comment by javierarce** [comment]
This looks good to me, nice job!

**Comment by jasmussen** [comment]
Apologies for coming back here after almost a month 😅 — but I much preferred the segmented control, and would want us to have that also for inline text colors:

Screenshot 2022-08-03 at 11 11 54

I don't dislike the tabs especially in global styles, though I still think the segmented control worked better there. The tab by virtue of going edge to edge acts as a separator or a new sub-section, when in fact it is contextual to the controls above and below both.

While I'd personally prefer us going back to the segmented control, I also don't mind trying this out for a bit. But we should at least change "Solid color" to just read "Solid". I'll spin up a PR.

**Comment by jasmussen** [comment]
I created #42918 as a followup to renaming, but learned that there are some awkward padding issues. I do see those raised in this thread, I wonder if that regressed? Would going back to the segmented control simplify some of this?

**Comment by mirka** [comment]
> Would going back to the segmented control simplify some of this?

Especially if we want the [full-width color button](https://github.com/WordPress/gutenberg/pull/41900#issuecomment-1203688470) to also happen, I'd say it would take more work to achieve everything correctly with the ToggleGroupControl.

The main hurdle with using the ToggleGroupControl as a `tabpanel` tab selector (i.e. `tablist`) is that ToggleGroupControl is semantically a `radio` right now. So the HTML would be very wrong. We would resolve this either by making a new TabPanel component, or add a design variant to the existing TabPanel component. (This is the "more work" part.)

I'm not entirely convinced that a radio control and a tab panel should look the same (e.g. what happens if you need to put a radio control inside the tab panel?), but if that's the direction we're going, it is technically possible.

**Comment by jasmussen** [comment]
> The main hurdle with using the ToggleGroupControl as a tabpanel tab selector (i.e. tablist) is that ToggleGroupControl is semantically a radio right now. So the HTML would be very wrong. We would resolve this either by making a new TabPanel component, or add a design variant to the existing TabPanel component. (This is the "more work" part.)

Just to be sure, that means when we had the ToggleGroupControl for solid/gradient instead of the new tab-group, the HTML was incorrect?

**Comment by mirka** [comment]
> Just to be sure, that means when we had the ToggleGroupControl for solid/gradient instead of the new tab-group, the HTML was incorrect?

Exactly. HTML-wise, that was just a radio control preceding the rest of the color controls. No hierarchy or indication that it was a tab panel switcher.

**Comment by jasmussen** [comment]
> Exactly. HTML-wise, that was just a radio control preceding the rest of the color controls. No hierarchy or indication that it was a tab panel switcher.

Thank you for explaining. In various forms isn't it reasonable enough to have a radio group contextually affect subsequent options? I don't think of the segmented control (ToggleGroupControl) as being conceptually related to a tab panel switcher, so unless there's some semantic reason not to use it, I'm unsure it's a problem. But I'm happy to learn here if I'm wrong.

**Comment by jameskoster** [comment]
Radios don't make sense in some flows, the inline text color options your shared in this thread are a good example:



You don't select text _or_ background, you can select both.

**Comment by jasmussen** [comment]
In that case tabs are more appropriate, indeed, but I feel the original point for customizing a single color still stands. It's not a very very strong opinion — things look okay. It's just not clear to me that the tabs are better for the color panel.

**Comment by jameskoster** [comment]
There may be an argument to support both, dependent on situation. In the example above, and when editing palettes in global styles, I think tabs make more sense. When applying a background color (solid vs gradient) to a specific block, radios / segmented control may be preferable.

**Pull request merged**
